### PR TITLE
Improve UTCTime encoding

### PR DESCRIPTION
**Pull request by bgamari**
The encodings given by the CBOR RFC are either inefficient (a
ISO8601 string) or imprecise (offset from the epoch in seconds).
Here we implement the extended time RFC (currently an [Internet Draft](https://tools.ietf.org/html/draft-bormann-cbor-time-tag-00#section-3.3)) which provides a greatly improved encoding.

Currently the draft lacks an assigned tag, so for the time being I've stolen `0x2221feca` (chosen at random via roll of a very large dice). I've started the IANA process for getting a tag assigned while the RFC process takes its course. Unfortunately, eventually updating the library to reflect the assigned tag will be a breaking change.

## Criterion Results

Benchmark | Before | After
------------|------------|-------
serialise UTCTime | 969.9 μs | 65.88 μs
deserialise UTCTime | 1.316 ms | 112.0 μs

### CODE REVIEW

File: Data/Binary/Serialise/CBOR/Class.hs [hunk 1]
```diff
@@ -1023,7 +1034,25 @@ instance Serialise UTCTime where
...
+            TypeInteger -> utcFromIntegral <$> decodeInteger
+          tt <- peekTokenType
+          picosecs <- case tt of
+            TypeUInt    -> utcFromIntegral <$> decodeWord
+            TypeUInt64  -> utcFromIntegral <$> decodeWord64
+            TypeNInt    -> utcFromIntegral <$> decodeInt
+            TypeNInt64  -> utcFromIntegral <$> decodeInt64
+            TypeInteger -> utcFromIntegral <$> decodeInteger
+          return (UTCTime (fromEnum days) (picosecondsToDiffTime picosecs))
+        _ -> fail "Expected timestamp (tag 0, 1 or 2)"
```
Comment (axman6) [inline Data/Binary/Serialise/CBOR/Class.hs]: "tag 0, 1 or 40" no?

Comment (axman6) [inline Data/Binary/Serialise/CBOR/Class.hs]: Also, `diffTimeToPicoseconds` and `picosecondsToDiffTime` need to be imported from `Data.Time` according to Travis...

Comment (bgamari) [inline Data/Binary/Serialise/CBOR/Class.hs]: Good call.

**Comment by bgamari** [comment]
> We can't change the existing instance anyway without breaking existing code, so this needs to be refactored anyway.

@thoughtpolice, I don't see how this would break existing code. We can read all of the encodings that we could previously. The only foreseeable issue is someone trying to read a file with a tag 40 encoding with an older version of the library. However, seeing as this isn't released yet I really don't see this as a problem. Afterall, they merely need to recompile their project against a newer `binary-serialise-cbor` to regain compatibility.

**Comment by bgamari** [comment]
@thoughtpolice ping?

File: Data/Binary/Serialise/CBOR/Class.hs (hunk 1 above)
Comment (axman6) [inline Data/Binary/Serialise/CBOR/Class.hs]: Pretty sure tag 2 shouldn't be there in the error message (at least, it doesn't look like the code supports a tag of 2).

Comment (axman6) [inline Data/Binary/Serialise/CBOR/Class.hs]: (unless we're supporting BigNum as a representation of time too)

**Comment by sopvop** [comment]
What about users who use CBOR to communicate with non haskell programms?

**Comment by bgamari** [comment]
Ahh, I neglected to mention that I am working to get the encoding proposed here (or something like it) granted an IANA-sanctioned tag. Moreover, the IETF has created a working group which will be taking CBOR to a full Internet standard; there is a fair chance that we will be able to get this encoding in that specification.

**Comment by thoughtpolice** [comment]
@bgamari Ah, no, you're correct. I misread the tag, so this is absolutely my fault because I'm an idiot.

The `docs` file should actually go in the Tutorial module or something I think, though; just having the documentation on the instance itself is rather out of line. But otherwise yeah, this seems OK.

**Comment by axman6** [comment]
I think my comments may have got lost/not seen because some of the code changed, but:

> Pretty sure tag 2 shouldn't be there in the error message (at least, it doesn't look like the code supports a tag of 2, unless we're supporting BigNum as a representation of time too).

**Comment by bgamari** [comment]
> I think my comments may have got lost/not seen because some of the code changed, but:

Thanks for the reminder, @axman6.

**Comment by bgamari** [comment]
For the record, we are currently in the final stages of submitting a final encoding for consideration by IETF. The final version is quite a bit different then what I have implemented here; I'll update this branch shortly.

**Comment by lfairy** [comment]
I found a draft of the proposal here: https://tools.ietf.org/html/draft-bormann-cbor-time-tag-00

Not sure how up to date it is, though.

**Comment by thoughtpolice** [comment]
Yes, it was published on the 13th of March (look below the author field) and the draft expires before September 23rd (a period during which revisions can be sent in and before another one has to be made). So I believe this has the differences Ben mentioned earlier (but he hasn't updated this PR)

(The IETF draft setup is a bit weird like that :)

**Comment by axman6** [comment]
It looks like that proposal has grown _a lot_ from the initial proposal here, is there a good reason for it to be so complex? Also reading the RFC, it would be good if there were examples of all the proposed structures. (making the comment here because I have no idea how actually RFC comments are supposed to work!)

**Comment by dcoutts** [comment]
> It looks like that proposal has grown a lot from the initial proposal here, is there a good reason for it to be so complex?

@axman6 You're of course welcome to contribute your feedback to the RFC process. I'd guess it's other feedback from other people about things they want.

But note that (unless I've missed something) we don't need to make use of all the flexibility available. It's worth looking specifically (and e.g. documenting here for the rest of us to review) what of the available options we would use, what that encoding would be exactly and how big it would be.

**Comment by bgamari** [comment]
Alex Mason <notifications@github.com> writes:

> It looks like that proposal has grown _a lot_ from the initial
> proposal here, is there a good reason for it to be so complex? Also
> reading the RFC, it would be good if there were examples of all the
> proposed structures. (making the comment here because I have no idea
> how actually RFC comments are supposed to work!)
>

This proposal isn't Haskell-centric; I'm merely one author. The proposal
is intended to address the general inadequacy of the date/time encodings
laid out in the original CBOR specification so yes, it is a bit more
complex than we strictly need.

**Comment by bgamari** [comment]
This should be ready for review.

File: src/Data/Binary/Serialise/CBOR/Class.hs [hunk 2]
```diff
@@ -1296,10 +1297,24 @@ instance Serialise TypeRep where
...
+        <> encodeMapLen 2
+        <> encodeWord 1 <> encodeInt64 secs
+        <> encodeInt (-12) <> encodeWord64 psecs
+      where
+        (secs, frac) = case properFraction $ utcTimeToPOSIXSeconds t of
+                         -- fractional part must be positive
+                         (secs', frac')
+                           | frac' < 0  -> (secs' - 1, frac' + 1)
+                           | otherwise -> (secs', frac')
+        psecs = round $ frac * 1000000000000
```
Comment (dcoutts) [inline src/Data/Binary/Serialise/CBOR/Class.hs]: looks fine for now, can optimise & iterate with criterion later.

**Comment by axman6** [comment]
Just confirming, this change doesn't implement the whole RFC, but only the bit necessary to efficiently encode UTCTimes?

I guess this should the optimised common case for encoding and decoding in Haskell, but maybe we should also provide a Times module which supports the whole RFC as both a proof of concept and to make life easier for those in the future need to interact with foreign code which does implement other parts/all of the RFC.

**Comment by bgamari** [comment]
> What do the criterion numbers look like now?

See PR summary.

**Comment by dcoutts** [comment]
No objection to providing greater support for the date/time RFC. PRs along those lines would be accepted.

Criterion numbers look good now.

Anything left to do or can we merge?

**Comment by bgamari** [comment]
We now have an official IANA sanctioned major tag, 1000!

I think this is ready to be merged.

**Pull request merged**